mlflow/sagemaker/__init__.py

```diff
@@ -287,40 +315,72 @@ def deploy(app_name, model_path, execution_role_arn=None, bucket=None, run_id=No
+                    " exists. If you want to update this application, deploy in '{mode_add}' or"
+                    " '{mode_replace}' mode.".format(
+                        application_name=app_name,
+                        mode_create=DEPLOYMENT_MODE_CREATE,
+                        mode_add=DEPLOYMENT_MODE_ADD,
+                        mode_replace=DEPLOYMENT_MODE_REPLACE)),
+                error_code=INVALID_PARAMETER_VALUE)
 
     if not image_url:
         image_url = _get_default_image_url(region_name=region_name)
-
     if not execution_role_arn:
         execution_role_arn = _get_assumed_role_arn()
-
     if not bucket:
         _logger.info("No model data bucket specified, using the default bucket")
         bucket = _get_default_s3_bucket(region_name)
 
-    model_s3_path = _upload_s3(
-        local_model_path=model_path, bucket=bucket, prefix=s3_bucket_prefix)
-    _deploy(role=execution_role_arn,
-            image_url=image_url,
-            app_name=app_name,
-            model_s3_path=model_s3_path,
-            run_id=run_id,
-            region_name=region_name,
-            mode=mode,
-            archive=archive,
-            instance_type=instance_type,
-            instance_count=instance_count,
-            vpc_config=vpc_config,
-            flavor=flavor)
+    model_s3_path = _upload_s3(local_model_path=model_path,
+                               bucket=bucket,
+                               prefix=s3_bucket_prefix,
+                               region_name=region_name,
+                               s3_client=s3_client)
+    if endpoint_exists:
+        deployment_operation = _update_sagemaker_endpoint(
+                endpoint_name=app_name, image_url=image_url, model_s3_path=model_s3_path,
+                run_id=run_id, flavor=flavor, instance_type=instance_type,
+                instance_count=instance_count, vpc_config=vpc_config, mode=mode,
+                role=execution_role_arn, sage_client=sage_client, s3_client=s3_client)
+    else:
+        deployment_operation = _create_sagemaker_endpoint(
+                endpoint_name=app_name, image_url=image_url, model_s3_path=model_s3_path,
+                run_id=run_id, flavor=flavor, instance_type=instance_type,
+                instance_count=instance_count, vpc_config=vpc_config, role=execution_role_arn,
+                sage_client=sage_client)
+
+    if synchronous:
+        _logger.info("Waiting for the deployment operation to complete...")
+        operation_status = deployment_operation.await_completion(timeout_seconds=timeout_seconds)
+        if operation_status.state != _SageMakerOperationStatus.STATE_SUCCEEDED:
+            raise MlflowException(
+                "The deployment operation failed with the following error message:"
+                " \"{error_message}\"".format(error_message=operation_status.message))
+        elif not archive and mode is not DEPLOYMENT_MODE_CREATE:
```

Comment:
Maybe instead of this if statement (which makes assumptions about the implementation of create's cleanup), we can simply not have a logger.info here and assume the cleanup method will log as appropriate?

Reply:
Good call. Removed the additional condition and moved the "Cleaning up..." log into the `cleanup_fn` function defined in the `_update_sagemaker_endpoint` method. I also moved the log statement in `delete` into the `delete#cleanup_fn` method.